### PR TITLE
Adding scale-up/down script

### PR DESCRIPTION
**Pull request by anshuvajpayee**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: samples/docdb-scaling/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+   - pymongo 4.4+
+   - If not installed - "$ pip3 install pymongo"
+ - AWS SDK for Python - tested versions 
+   - boto3 1.28+
+   - If not installed - "$ pip3 install boto3"
+
+
+## Using the Scale Up/Down Tool
+`python3 scaleupdown.py  --cluster-id <cluster-id> --target-instance-type <target-instance-type>  `
+- Run on any AWS Cloud9 or Bastion host(Jump host) where aws cli is configured.
```
Comment (tmcallaghan) [inline samples/docdb-scaling/README.md:22]: Why does the aws cli need to be present?

File: samples/docdb-scaling/scaleupdown.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        return True
+    else:
+        return False
+
+def checkClusterHealth(cluster_id,mins=15):
+    count=0
+    while not isClusterHealthy(cluster_id=cluster_id):
+        count=count+1
+        if count==mins:
+            LOGGER.error("Cluster status is not healthy after " + str(mins) +  " mins. Please check the cluster health .Exiting.. ")
```
Comment (tmcallaghan) [inline samples/docdb-scaling/scaleupdown.py:111]: What is our guidance here? Needs to be more than "Please check the cluster health".

File: samples/docdb-scaling/scaleupdown.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        sys.exit(1)
+        return False 
+    
+    num_instances=getInstanceStatus(cluster_id)["num_instances"]
+    oinstances=list((getClusterInstances(cluster_id)).keys())
+    
+    if num_instances > 8:
+        LOGGER.error("Number of instances can not be more than 8 for this script.")
+        sys.exit(1)
+        return False
```
Comment (tmcallaghan) [inline samples/docdb-scaling/scaleupdown.py:162]: dead code, can't get here, remove.

File: samples/docdb-scaling/scaleupdown.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        sys.exit(1)
+   
+    if target_instance_check and checkExistingInstanceClass(cluster_id=cluster_id,itarget=itarget):
+        LOGGER.error("Cluster already has one of the instance with target instance type.If you want to still continue,please run with --ignore-instance-check.Exiting..")
+        sys.exit(1)
+
+    if not isClusterHealthy(cluster_id):
+        LOGGER.error("Cluster is not healthy. Exiting..")
+        sys.exit(1)
+        return False
```
Comment (tmcallaghan) [inline samples/docdb-scaling/scaleupdown.py:154]: dead code, can't get here, remove.

File: samples/docdb-scaling/scaleupdown.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        sys.exit(1)
+    
+    failover_target=ninstances[0]
+    old_writer=getWriterInstance(cluster_id=cluster_id)
+    LOGGER.info("Failover target - " + failover_target)
+    
+    ## Perform failover 
+    perfromFailover(cluster_id=cluster_id,failover_target=failover_target)
+    LOGGER.info("checking failover status")
+    for count in range(4):
```
Comment (tmcallaghan) [inline samples/docdb-scaling/scaleupdown.py:182]: Why are we checking 4 times? What if it takes longer than 2 minutes?

File: samples/docdb-scaling/scaleupdown.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
+import boto3
+import json
```
Comment (mihaialdoiu) [inline samples/docdb-scaling/scaleupdown.py:2]: I don't see json and os modules being used, remove if that's the case.

File: samples/docdb-scaling/scaleupdown.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            LOGGER.error("Cluster status is not healthy after " + str(mins) +  " mins. Please check the cluster health .Exiting.. ")
+            return False
+        LOGGER.info("will be checking cluster health in 60 seconds")
+        time.sleep(60)
+    return True
+
+def main():
+    
+    parser = argparse.ArgumentParser(description='Amazon DocumentDB Cluster Scale up/down Tool.')
+    parser.add_argument('--cluster-id',
```
Comment (mihaialdoiu) [inline samples/docdb-scaling/scaleupdown.py:120]: I'm thinking that we want to validate arguments passed to the script. Ensure that the cluster ID and target instance type are provided as expected.

**Review by mihaialdoiu — COMMENTED** [review]
General comments:
- Add comments and the purpose and usage of functions or at least important code sections
- You're checking if the cluster status is available in multiple places. You can refactor it into a separate function.
- Don't forget error handling, as Tim said
- Consider using Boto3 waiters instead of manual sleep and polling -https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/docdb.html#waiters

File: samples/docdb-scaling/scaleupdown.py (hunk 3 above)
Comment (anshuvajpayee) [reply samples/docdb-scaling/scaleupdown.py:162]: fixed

File: samples/docdb-scaling/scaleupdown.py (hunk 6 above)
Comment (anshuvajpayee) [reply samples/docdb-scaling/scaleupdown.py:2]: removed

File: samples/docdb-scaling/scaleupdown.py (hunk 7 above)
Comment (anshuvajpayee) [reply samples/docdb-scaling/scaleupdown.py:120]: that I am checking in code.

File: samples/docdb-scaling/scaleupdown.py (hunk 5 above)
Comment (anshuvajpayee) [reply samples/docdb-scaling/scaleupdown.py:182]: failover generally  completes in  within 30 seconds.  I am checking here it for 2 mins.

File: samples/docdb-scaling/README.md (hunk 1 above)
Comment (anshuvajpayee) [reply samples/docdb-scaling/README.md:22]: Tool will require credentials for Boto3. I updated recommended way to use it.

File: samples/docdb-scaling/scaleupdown.py (hunk 2 above)
Comment (anshuvajpayee) [reply samples/docdb-scaling/scaleupdown.py:111]: They need to make sure cluster is in healthy state before running script. Added details.

**Comment by anshuvajpayee** [comment]
fixed all the issues and push new code.

**Review by tmcallaghan — APPROVED** [review]
Approved.

**Pull request merged**